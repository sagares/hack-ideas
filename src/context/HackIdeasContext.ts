import { createContext } from "react";
import { IdeasArray } from "../HackIdea.types";

type Employee = {
  empId: string;
};

type HackIdeasContextType = {
  currentUser: Employee;
  logout: () => void;
  setCurrentUser: (employee: Employee) => void;
  ideas: IdeasArray;
  setIdeas: (ideas: IdeasArray) => void;
};

const HackIdeasContext = createContext<Partial<HackIdeasContextType>>({});

export { HackIdeasContext };
