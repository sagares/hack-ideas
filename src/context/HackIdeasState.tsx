import React, { useReducer } from "react";
import { Employee, IdeasArray } from "../HackIdea.types";
import { HackIdeasContext } from "./HackIdeasContext";
import HackIdeasReducer from "./HackIdeasReducer";
import { LOG_OUT, SET_CURRENT_USER, SET_IDEAS } from "./Actions";

const HackIdeasState = ({ children }: any) => {
  const initialState = {
    currentUser: null,
    ideas: null,
  };

  const [state, dispatch] = useReducer(HackIdeasReducer, initialState);
  const setCurrentUser = (employee: Employee) => {
    dispatch({
      type: SET_CURRENT_USER,
      payload: employee,
    });
  };

  const setIdeas = (ideas: IdeasArray) => {
    dispatch({
      type: SET_IDEAS,
      payload: ideas,
    });
  };

  const logout = () => {
    dispatch({
      type: LOG_OUT,
      payload: initialState
    });
  };

  return (
    <HackIdeasContext.Provider
      value={{
        currentUser: state.currentUser,
        ideas: state.ideas,
        logout,
        setCurrentUser,
        setIdeas
      }}
    >
      {children}
    </HackIdeasContext.Provider>
  );
};

export { HackIdeasState };
