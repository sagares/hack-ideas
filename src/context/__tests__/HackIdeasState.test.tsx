/**
 * @jest-environment jsdom
 */
import React from "react";
import { fireEvent, render } from "@testing-library/react";

import { HackIdeasState } from "../HackIdeasState";
import { HackIdeasContext } from "../HackIdeasContext";

describe("Hack Ideas State", () => {
  let element, component, HackIdeasStateComponent, dispatch;

  beforeEach(() => {
    HackIdeasStateComponent = () => (
      <HackIdeasState>
        <HackIdeasContext.Consumer>
          {(context) => (
            <>
              <div
                className="set-current-user"
                onClick={() => {
                  context.setCurrentUser({ empId: "123456" });
                }}
              ></div>
              <div
                className="set-ideas"
                onClick={() => {
                  context.setIdeas({
                    idea1: {
                      name: "idea1",
                      description: "idea2",
                      id: "idea1",
                      tags: [],
                      submittedBy: "123456",
                      submittedOn: 1000,
                    },
                  });
                }}
              ></div>
              <div
                className="log-out"
                onClick={() => {
                  context.logout();
                }}
              ></div>
              <div className="current-user">{context.currentUser?.empId}</div>
              <div className="ideas">{context.ideas && context.ideas["idea1"].name}</div>
            </>
          )}
        </HackIdeasContext.Consumer>
      </HackIdeasState>
    );

    element = document.createElement("div");

    component = render(<HackIdeasStateComponent />, {
      container: document.body.appendChild(element),
    });
  });

  test("renders hack ideas state", () => {
    expect(component.container).toMatchSnapshot();
  });

  test("sets current user", () => {
    const setCurrentUser = element.querySelector(".set-current-user");
    fireEvent.click(setCurrentUser);

    expect(element.querySelector(".current-user").innerHTML).toEqual("123456");
  });

  test("sets ideas", () => {
    const setIdeas = element.querySelector(".set-ideas");
    fireEvent.click(setIdeas);

    expect(element.querySelector(".ideas").innerHTML).toEqual("idea1");
  });

  test("logs out current user", () => {
    const logout = element.querySelector(".log-out");
    fireEvent.click(logout);

    expect(element.querySelector(".current-user").innerHTML).toEqual("");
    expect(element.querySelector(".ideas").innerHTML).toEqual("");
  });
});
