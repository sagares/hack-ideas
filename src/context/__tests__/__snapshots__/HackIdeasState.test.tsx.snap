// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hack Ideas State renders hack ideas state 1`] = `
<div>
  <div
    class="set-current-user"
  />
  <div
    class="set-ideas"
  />
  <div
    class="log-out"
  />
  <div
    class="current-user"
  />
  <div
    class="ideas"
  />
</div>
`;
